sam0 DMA and UART bugfixes

rocky134 opened the pull request:
Several bugfixes related to the sam0 platform when using async UART with DMA transfers.

4a4ae4f05378916ad87c5958f8850c97e481a2f7 fixes the issue in which the sam0 uart driver is loaded in PRE_KERNEL_1 as CONFIG_KERNEL_INIT_PRIORITY_DEVICE (default level 50) but when utilizing CONFIG_UART_ASYNC_API the dma device is not ready during init since it is only initialized in POST_KERNEL.
Changing the sam0 DMA driver to load during PRE_KERNEL_1 with CONFIG_KERNEL_INIT_PRIORITY_DEFAULT (default level 40) allows DMA to be ready.

c29519fd06f4d06b7000d7c8359bc48e5d37fc98 fixes the issue building without CONFIG_UART_INTERRUPT_DRIVEN causes a build failure since uart_sam0_err_check is guarded by the above define but .err_check is not.

e116aefcf3c58776dc4110a958c3dc6eabaf67b3 fixes the issue when using the async driver with the UART_TX_DONE event not occurring when TX is actually done.
The sam0 device generates a DMA interrupt when the DMA transfer from the TX ram buffer to the SERCOM DATA register is completed **not** when the data is finished being sent.
Therefore when the user callback is expecting all TX data to be completed the transfer is still ongoing and if UART_TX_DONE should trigger an external hardware event such as disabling a RS485 transceiver, data corruption will occur.

The following was tested with an ATSAMD21 device.

ElectronFlux committed:
drivers: dma: dma_sam0: change sam0 DMA to higher priority init.

fixes peripheral drivers such as async uart that rely on dma being
ready failing because dma is not initialized yet.

Signed-off-by: Ron Smith <rockyowl171@gmail.com>

ElectronFlux committed:
drivers: serial: uart_sam0: move err_check methodout of if guard

Fixes a compile error for the err_check function not being found if
if CONFIG_UART_INTERRUPT_DRIVEN is not enabled.

Signed-off-by: Ron Smith <rockyowl171@gmail.com>

ElectronFlux committed:
drivers: serial: uart_sam0: Fix async tx done event triggering to early.

uart_sam0_dma_tx_done callback triggers when the last byte
is transferred from the tx sram buffer to the sercom DATA register.
However the byte has yet to be transmitted completely which can lead to
incorrect event handling if UART_TX_DONE is expected to signal
the end of transmission.

Signed-off-by: Ron Smith <rockyowl171@gmail.com>

nandojve commented:
I changed from:
https://github.com/zephyrproject-rtos/zephyr/blob/fcfe597e5dadb96ecef5d6e8c82ffc99b5f4d59c/tests/drivers/uart/uart_async_api/boards/atsamr21_xpro.overlay#L1-L15

to:
```
&dmac {
	status = "okay";
};

&sercom0 {
	/* Configure DMA channels for async operation */
	dmas = <&dmac 0 1>, <&dmac 1 2>;
	dma-names = "rx", "tx";
};

&sercom3 {
	status = "okay";
	compatible = "atmel,sam0-uart";
	current-speed = <115200>;

	/* Internally loop-back TX and RX on PAD0 */
	rxpo = <0>;
	txpo = <0>;
	pinctrl-0 = <&pa17d_sercom3_pad1 &pa16d_sercom3_pad0>;

	/* Configure DMA channels for async operation */
	dmas = <&dmac 0 7>, <&dmac 1 8>;
	dma-names = "rx", "tx";
};
```

I'm still not get characters at console. I noted that a hard fault happen but I couldn' t get data on console.

@rocky134 , how did you conduct your test? Can describe it? I wondering if missing something.

Note:

1) This PR is passing CI because Zephyr is skipping SAM0 boards at `tests/drivers/uart/uart_async_api/testcase.yaml`.
2) Current pinctrl code requires pin definitions even for loopback interfaces. It is something that may require improvements.

nandojve commented:
I conduct more tests and noted that sam0 maybe is broke at main.

I restored to e49c7f1780 (random commit - 4 weeks ago - need bisect) and I got below:

```
*** Booting Zephyr OS build zephyr-v2.6.0-1658-ga7cdaf9ebf08  ***
Running test suite uart_async_test
===================================================================
START - test_single_read_setup
 PASS - test_single_read_setup in 0.1 seconds
===================================================================
START - test_single_read

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:82: test_single_read: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
*** Booting Zephyr OS build zephyr-v2.6.0-1658-ga7cdaf9ebf08  ***
Running test suite uart_async_test
===================================================================
START - test_single_read_setup
 PASS - test_single_read_setup in 0.1 seconds
===================================================================
START - test_single_read

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:82: test_single_read: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
```

small note for reference: c8cca16c20 - last commit that above test pass before regression (without rocky134 changes).

@rocky134, can you try run above tests and report to us?

BTW, thank you for your PR.

rocky134 commented:
> @rocky134 , how did you conduct your test? Can describe it? I wondering if missing something.

I first discovered this when data corruption was occurring when handling UART_TX_DONE in the callback I would drive a GPIO low to signify transmit complete to a RS485 half duplex transceiver.  After noticing the last byte would always be corrupted, under an oscilloscope I found that the GPIO would be driven low half way through the last byte being transmitted, to which I traced back the UART_TX_DONE event coming from the DMA transfer complete, not the tx completed.

Once I made the changes data corruption stopped occurring transmitting and receiving still worked, and I verified with an oscilloscope and system view that the sequence of events occurred in the right order. Then I was satisfied that the problem was fixed.
I added the following before taking the systemview sample.
```
--- a/drivers/serial/uart_sam0.c
+++ b/drivers/serial/uart_sam0.c
@@ -14,6 +14,7 @@
 #include <drivers/uart.h>
 #include <drivers/dma.h>
 #include <string.h>
+#include "SEGGER_SYSVIEW.h"
 
 #ifndef SERCOM_USART_CTRLA_MODE_USART_INT_CLK
 #define SERCOM_USART_CTRLA_MODE_USART_INT_CLK SERCOM_USART_CTRLA_MODE(0x1)
@@ -144,6 +145,7 @@ static void uart_sam0_dma_tx_done(const struct device *dma_dev, void *arg,
        SercomUsart * const regs = cfg->regs;

        regs->INTENSET.reg = SERCOM_USART_INTENSET_TXC;
+       SEGGER_SYSVIEW_MarkStart(0x4);
 }

 static int uart_sam0_tx_halt(struct uart_sam0_dev_data *dev_data)
@@ -673,6 +675,7 @@ static void uart_sam0_isr(const struct device *dev)

        if (dev_data->tx_len && regs->INTFLAG.bit.TXC) {
                regs->INTENCLR.reg = SERCOM_USART_INTENCLR_TXC;
+               SEGGER_SYSVIEW_MarkStop(0x4);

                k_work_cancel_delayable(&dev_data->tx_timeout_work);
```
![image](https://user-images.githubusercontent.com/34631288/131258051-2eff0fb3-c056-463c-8baa-ff86b053cded.png)




>I'm still not get characters at console. I noted that a hard fault happen but I couldn't' t get data on console.

yes, I noticed that too on the most recent main branch, my local branch I was working in was up to f3435b75efbfd7614ff873f8bfa90a031d6d27e8. One issue I had that was new was in pinmux.c muxa would not be ready since they're at the same initialization level. changing pinmux.c to init at PRE_KERNEL_2 fixed that problem.

> @rocky134, can you try run above tests and report to us?

I ran the tests in zephyr/tests/drivers/uart/uart_async_api with the following device tree fragment,
```
&sercom3 {
	status = "okay";
	compatible = "atmel,sam0-uart";
	current-speed = <115200>;
	rxpo = <0>;
	txpo = <0>;
	dmas = <&dmac 0 0x08>, <&dmac 1 0x07>;
	dma-names = "tx", "rx";

	pinctrl-0 = <&pa23c_sercom3_pad1>, // rx pin
				<&pa22c_sercom3_pad0>; // tx pin
};
```
(I am using a custom board based off the atsamd21_xpro but using the ATSAMD21G17A instead. )
however I was not able to get the tests to pass either, same as your errors. both with and without my changes. I also tried changing 
```
rxpo = <1>;
```
and physically soldered a wire between RX and TX but that did not pass either.
I tested the above on both the current main branch commit fcfe597e5dadb96ecef5d6e8c82ffc99b5f4d59c with my merge changes and a previous commit without any changes but I was not able to get the tests to pass.

nandojve commented:
Hi @rocky134,

The #38089 fixes init hard fault. Thank you for sharing solution.

rocky134 commented:
It seems a bit strange how the line in this PR [R144](https://github.com/zephyrproject-rtos/zephyr/commit/627146e8bbe38e75e4b5a37b3e66bb58a251da93#diff-559097f3a45b99a62a8e375a4a5b7ffe6809e0a636f3cc4449a12116f251373aR144) causes a coding style compliance fail when other uses of this declaration at [L249](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_sam0.c#L249), [L327](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_sam0.c#L327) and [L1057](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_sam0.c#L1057) were merged without compliance failures.

cfriedt commented:
@rocky134 - are there any issues that can be linked to this PR? I believe the compliance failure may be a false positive. @galak, can you confirm? 

@nandojve, can you please review?

rocky134 commented:
> are there any issues that can be linked to this PR?

As far as I know, there are not any open issues related specifically to this PR however it is still an issue with how async serial is being handled that needs to be fixed. As stated in the original post, handling UART_TX_DONE expecting transmission to be completed when it actually isn't yields data loss when used to disable the DE pin of a half duplex transceiver, similar to issue #38264.

I chose to create a PR with the fix rather than just open an issue since I already fixed the issue after discovering it. Creating this PR was an afterthought to share my fix with others.

cfriedt commented:
@rocky134 - compliance failure is fixed, LGTM, but will still need one other review. @nandojve ?

cfriedt reviewed (approved):
LGTM

nandojve commented:
> @rocky134 - compliance failure is fixed, LGTM, but will still need one other review. @nandojve ?

Hi @cfriedt , I'll check this weekend.

nandojve commented:
@rocky134 ,

I was wondering if you could add below changes as two newer commits. One describes how proper configure DMA channels. The other one configure samr21 to pass at `tests/drivers/uart/uart_async_api`. The error happen because cells `channel` must be unique when configuring multiple peripherals. I mean, SERCOM0 uses 2 channels [0,1] and SERCOM3 should use 2 different channels [10, 11], for instance.

I think your PR fixes #35721 but need someone run tests on some SAME5x board CC: @stephanosio @arvinf ?

```diff
diff --git a/dts/bindings/dma/atmel,sam0-dmac.yaml b/dts/bindings/dma/atmel,sam0-dmac.yaml
index 5260e3624f..739065dbb3 100644
--- a/dts/bindings/dma/atmel,sam0-dmac.yaml
+++ b/dts/bindings/dma/atmel,sam0-dmac.yaml
@@ -14,6 +14,26 @@ properties:
     "#dma-cells":
       const: 2
 
+# #dma-cells : Must be <2>.
+# The 1st cell specifies the DMAC channel to be used for the data transfer.
+#   This channel should be unique between all peripherals that are using the
+#   DMAC instance.
+# The 2nd cell define the peripheral trigger which is source of the transfer.
+#   For details on trigger selection and trigger modes, refer to
+#   "Transfer Triggers and Actions".
+#   See the SoC's reference manual for all the supported request sources.
+#
+# Example of devicetree dma channel configuration:
+#
+# &sercom3 {
+#     /* Configure DMA channels for async operation */
+#     dmas = <&dmac 10 7>, <&dmac 11 8>;
+#     dma-names = "rx", "tx";
+# };
+#
+# In above fragment 10 and 11 represents the different channels used to
+# transfer data between peripheral and ram. The numbers 7/8 are, for instance,
+# uart_rx/tx peripheral trigger for sercom3.
 dma-cells:
   - channel
   - trigsrc
diff --git a/tests/drivers/uart/uart_async_api/boards/atsamr21_xpro.overlay b/tests/drivers/uart/uart_async_api/boards/atsamr21_xpro.overlay
index 8d39931188..e4df1a1fda 100644
--- a/tests/drivers/uart/uart_async_api/boards/atsamr21_xpro.overlay
+++ b/tests/drivers/uart/uart_async_api/boards/atsamr21_xpro.overlay
@@ -1,5 +1,15 @@
 /* SPDX-License-Identifier: Apache-2.0 */
 
+&dmac {
+       status = "okay";
+};
+
+&sercom0 {
+       /* Configure DMA channels for async operation */
+       dmas = <&dmac 0 1>, <&dmac 1 2>;
+       dma-names = "rx", "tx";
+};
+
 &sercom3 {
        status = "okay";
        compatible = "atmel,sam0-uart";
@@ -9,7 +19,10 @@
        rxpo = <0>;
        txpo = <0>;
 
+       /* PAD0 must be configured to allow loop-back works */
+       pinctrl-0 = <&pa16d_sercom3_pad0>;
+
        /* Configure DMA channels for async operation */
-       dmas = <&dmac 0 7>, <&dmac 1 8>;
+       dmas = <&dmac 10 7>, <&dmac 11 8>;
        dma-names = "rx", "tx";
 };
```

rocky134 commented on drivers/serial/uart_sam0.c:
```diff
@@ -1126,8 +1126,9 @@ static const struct uart_driver_api uart_sam0_driver_api = {
-	.err_check = uart_sam0_err_check,
+
 #if CONFIG_UART_INTERRUPT_DRIVEN
+	.err_check = uart_sam0_err_check,
```
The error I get that prompted that fix is when building with CONFIG_UART_ASYNC_API  but not CONFIG_UART_INTERRUPT_DRIVEN.
Since the actual function uart_sam0_err_check is [guarded by the ifdef](https://github.com/zephyrproject-rtos/zephyr/blob/11df9acd12a86059fea2533a38e4067261ffcecc/drivers/serial/uart_sam0.c#L724) a compile error is raised as the function is not defined. If it must always pre present then it should be moved outside the `#ifdef CONFIG_UART_INTERRUPT_DRIVEN`

rocky134 replied:
since .err_check  is a required method, I moved the uart_sam0_err_check block outside of the if guard to allow for compiling when CONFIG_UART_INTERRUPT_DRIVEN is not used. aca350547b8ff15f7f8fbe187ee7e8d48139700b

ElectronFlux committed:
dts: atmel: sam0: Add clarification on dma-cells property

Add clarification to the dma-cells property explaining the purpose
of each cell.

Signed-off-by: Ron Smith <rockyowl171@gmail.com>

ElectronFlux committed:
tests: uart_async_api: Correct DMA channels on atsamr21_xpro board

This commit corrects the DMA channels for the asynchronous
UART API testing support on the SAM R21 Xplained Pro board.

Signed-off-by: Ron Smith <rockyowl171@gmail.com>

rocky134 commented:
> I was wondering if you could add below changes as two newer commits.

they have been added @nandojve

nandojve reviewed (approved):
Thank you so much @rocky134 ,

I hope I can see more PRs from you!


```
*** Booting Zephyr OS build v2.7.0-rc2-12-g614ec885b96a  ***
Running test suite uart_async_test
===================================================================
START - test_single_read_setup
 PASS - test_single_read_setup in 0.1 seconds
===================================================================
START - test_single_read
 PASS - test_single_read in 0.352 seconds
===================================================================
START - test_chained_read_setup
 PASS - test_chained_read_setup in 0.1 seconds
===================================================================
START - test_chained_read
 PASS - test_chained_read in 0.217 seconds
===================================================================
START - test_double_buffer_setup
 PASS - test_double_buffer_setup in 0.1 seconds
===================================================================
START - test_double_buffer
 PASS - test_double_buffer in 3.401 seconds
===================================================================
START - test_read_abort_setup
 PASS - test_read_abort_setup in 0.1 seconds
===================================================================
START - test_read_abort
 PASS - test_read_abort in 1.60 seconds
===================================================================
START - test_chained_write_setup
 PASS - test_chained_write_setup in 0.1 seconds
===================================================================
START - test_chained_write
 PASS - test_chained_write in 0.3 seconds
===================================================================
START - test_long_buffers_setup
 PASS - test_long_buffers_setup in 0.1 seconds
===================================================================
START - test_long_buffers
 PASS - test_long_buffers in 0.150 seconds
===================================================================
START - test_write_abort_setup
 PASS - test_write_abort_setup in 0.1 seconds
===================================================================
START - test_write_abort
 PASS - test_write_abort in 0.103 seconds
===================================================================
START - test_forever_timeout_setup
 PASS - test_forever_timeout_setup in 0.1 seconds
===================================================================
START - test_forever_timeout
 PASS - test_forever_timeout in 3.1 seconds
===================================================================
Test suite uart_async_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

rocky134 commented:
rebased PR to current main head per recommendation here https://github.com/zephyrproject-rtos/zephyr/issues/35721#issuecomment-922394760
